### PR TITLE
GCP Observability : Framework for detecting the environment

### PR DESCRIPTION
**Pull request by yashykt**
This code is not plumbed through yet, but it provides the core infrastructure needed to detect the proper GCP environment resources needed to set up the labels/attributes/resources for stats, tracing and logging.

Details on how the various environment resources are setup has been derived by looking at java's cloud logging library and OpenTelemetry's future plans. (Could be better explained in an offline review since some links are internal).

Requesting @veblush for a full review and @markdroth for a structural review.

### CODE REVIEW

**Review by veblush — APPROVED** [review]

File: src/cpp/ext/filters/census/environment_autodetect.cc
Comment (ctiller) [inline src/cpp/ext/filters/census/environment_autodetect.cc:241]: This seems like it would be simpler & clearer if it was a `flat_hash_map<string /* metadata_server_attribute */, string /* resource_attribute */>`

Comment (yashykt) [reply src/cpp/ext/filters/census/environment_autodetect.cc:241]: suggestion taken

File: src/cpp/ext/filters/census/environment_autodetect.cc
Comment (drfloob) [inline src/cpp/ext/filters/census/environment_autodetect.cc:223]: This can be replaced by `GetDefaultEventEngine()->Run([this]{ PollLoop(); });`. 

Note that with either a thread like this or using EventEngine::Run, we don't strictly have to worry about fork support since this is in C++ code, not the core library where Python requires a clean fork story, but it might be worth adding a comment that this is not fork-safe for anyone who wants fork support in C++. If you want to make this fork-safe, we can have a short design chat.

Comment (yashykt) [reply src/cpp/ext/filters/census/environment_autodetect.cc]: Suggestion accepted. Added comments to that affect when we actually care about fork support here.

File: src/cpp/ext/filters/census/environment_autodetect.cc
Comment (markdroth) [inline src/cpp/ext/filters/census/environment_autodetect.cc:78]: Since this code essentially already exists in the google-c2p resolver, I think we should move it into a common library instead of duplicating it.

File: src/cpp/ext/filters/census/environment_autodetect.cc
Comment (drfloob) [inline src/cpp/ext/filters/census/environment_autodetect.cc:229]: Note that this is potentially unsafe usage of `GetDefaultEventEngine`. It returns a `shared_ptr`, and will lazily create an instance if one does not already exist. If you do not save the engine locally, you risk having the engine created and destroyed on this line.

If you aren't certain the default engine's lifetime is guaranteed until your timer is complete, I'd recommend adding a `shared_ptr<EventEngine> engine_` member to `EnvironmentAutoDetectHelper` to cache it for the life of the object at least.

File: src/cpp/ext/filters/census/environment_autodetect.h
Comment (markdroth) [inline src/cpp/ext/filters/census/environment_autodetect.h:35]: Does this code actually belong in the census filter directory?  It seems to me that it's really a more general GCP-specific autodetection thing, not really anything specific to the census filter.  Maybe this should go in its own library somewhere?

Note that we also have some other existing code to detect GCP platform here, which we use in ALTS and in GoogleDefaultCreds:

https://github.com/grpc/grpc/blob/0ecc18ef0f51eb38880835db34f38f99980367e4/src/core/lib/security/credentials/alts/check_gcp_environment.h#L55

@apolcyn may have some thoughts about the best way to structure all of this code into a single combined library.

File: src/cpp/ext/filters/census/environment_autodetect.h
Comment (markdroth) [inline src/cpp/ext/filters/census/environment_autodetect.h:39]: Suggest making this an enum instead of a string.

File: src/cpp/ext/filters/census/environment_autodetect.h
Comment (markdroth) [inline src/cpp/ext/filters/census/environment_autodetect.h:47]: Splitting `Create()` and `Get()` into two different methods seems odd.  If there are multiple callers, how do any of them know which one is the first one?

Also, what happens if `Create()` is called twice, but with different project IDs?

File: src/cpp/ext/filters/census/environment_autodetect.cc
Comment (markdroth) [inline src/cpp/ext/filters/census/environment_autodetect.cc:413]: I generally consider `ReleasableMutexLock` to be a code smell.  And in this case, since you're invoking the callback asynchronously, it looks like you don't actually need to release the lock early in the first place.

Comment (yashykt) [reply src/cpp/ext/filters/census/environment_autodetect.cc]: Done. I was intending to reduce the critical region as much as I could, but the EventEngine::Run call is not so bad that we need to worry about it here.

File: src/cpp/ext/filters/census/environment_autodetect.cc
Comment (drfloob) [inline src/cpp/ext/filters/census/environment_autodetect.cc:126]: I don't see this constructor used with an initializer list, and I'm not sure you'd want to prevent that regardless, so removing `explicit` here seems the right thing to do.

File: src/core/ext/gcp/metadata_query.cc
Comment (drfloob) [inline src/core/ext/gcp/metadata_query.cc:89]: Small suggestion: guard the logging behind a trace flag. Trace flags can be set on by default, if that's your intention, but it will provide an easy toggle later if you want to disable this logging as on by default.

File: src/cpp/ext/filters/census/environment_autodetect.cc
Comment (drfloob) [inline src/cpp/ext/filters/census/environment_autodetect.cc:183]: The performance hit is probably negligible, but I wonder if this would be better written as a loop with a 100ms sleep, than a lambda-wrapped recursive function with a thread hop.

```
bool done = false;
while (!done) {
  // ...
}
Unref();
```

File: src/cpp/ext/filters/census/environment_autodetect.cc
Comment (yashykt) [reply src/cpp/ext/filters/census/environment_autodetect.cc]: Done

File: src/core/ext/gcp/metadata_query.cc
Comment (yashykt) [reply src/core/ext/gcp/metadata_query.cc:89]: Note that all these logs are error logs which we generally do not guard by Trace Flags. I think we always want to see these logs.

File: src/cpp/ext/filters/census/environment_autodetect.cc
Comment (yashykt) [reply src/cpp/ext/filters/census/environment_autodetect.cc:183]: Isn't it bad to do a blocking sleep in an event manager thread. That seems like something we definitely do not want to do. Also note that in this case, we really do not care about performance.

File: src/cpp/ext/filters/census/environment_autodetect.cc
Comment (yashykt) [reply src/cpp/ext/filters/census/environment_autodetect.cc:126]: Thanks!

File: src/cpp/ext/filters/census/environment_autodetect.h
Comment (yashykt) [reply src/cpp/ext/filters/census/environment_autodetect.h]: Moved it to src/cpp/ext/gcp/... Alex, let's chat on whether we'd like to move some of this in Core if it's reusable.
I'd like to not block this PR though.

File: src/core/ext/gcp/metadata_query.h
Comment (apolcyn) [inline src/core/ext/gcp/metadata_query.h:36]: The existing `MetadataQuery` object used by the google-c2p resolver is cancellable, and e.g. the resolver will cancel outstanding MD server queries during shutdown.

Looks like cancellation is removed from the revised API, is that intentional?

File: src/core/ext/gcp/metadata_query.cc
Comment (apolcyn) [inline src/core/ext/gcp/metadata_query.cc:76]: 1 second seems like a fairly short default timeout. The google-c2p resolver currently uses 10 seconds. (I'm not sure what the "right" value is, but we may want to give some slack e.g. for packet loss b/t the client and MD server).

Maybe, make this timeout configurable?

File: src/core/ext/gcp/metadata_query.cc
Comment (apolcyn) [inline src/core/ext/gcp/metadata_query.cc:92]: We shouldn't log ERROR here, since there are some queries that are performed as "probes" for which we may or may not get a 200 response, but we can handle both cases.

E.g. the google-c2p resolver queries the MD server to see if the VM has an IPv6 address, to decide whether to prefer v4 or v6.

If we want to add extra error message context, suggest adding it to the error and letting the caller decide to log or not.

**Review by apolcyn — CHANGES_REQUESTED** [review]

File: src/core/ext/gcp/metadata_query.cc
Comment (drfloob) [reply src/core/ext/gcp/metadata_query.cc:89]: Some are DEBUG, and the first one in the Poll loop is an ERROR and it maybe should not be? Your call though.

File: src/core/ext/gcp/metadata_query.h
Comment (yashykt) [reply src/core/ext/gcp/metadata_query.h:36]: MetadataQuery requests are intended to be quick enough that I don't think that the cancellation matters so much but it is also trivial enough to add, so I don't mind adding it.

File: src/core/ext/gcp/metadata_query.cc
Comment (yashykt) [reply src/core/ext/gcp/metadata_query.cc]: Given that Alex had similar concerns, guarded it with TraceFlags

File: src/core/ext/gcp/metadata_query.cc
Comment (yashykt) [reply src/core/ext/gcp/metadata_query.cc]: Done!

File: src/core/ext/gcp/metadata_query.h
Comment (yashykt) [reply src/core/ext/gcp/metadata_query.h]: Done

File: src/core/ext/gcp/metadata_query.cc
Comment (apolcyn) [inline src/core/ext/gcp/metadata_query.cc:76]: nit: stale comment now that timeout is a param?

File: src/core/ext/gcp/metadata_query.cc
Comment (apolcyn) [inline src/core/ext/gcp/metadata_query.cc:102]: nit: as long as this tracer flag is enabled, it will probably be helpful to log all MD server query results, and the HTTP status will probably be useful to include in the log too.

e.g. suggest replacing:

```
if (!error.ok()) {
  ...
} else if (self->response_.status != 200) {
  ...
}
```

with just one log at the top:

```
if (GRPC_TRACE_FLAG_ENABLED(grpc_metadata_query_trace)) {
    gpr_log(GPR_INFO, "MetadataServer Query for %s: http status: %d, error: %s",
            self->attribute_.c_str(), self->response_.status, StatusToString(error).c_str());
}
```

File: src/core/ext/gcp/metadata_query.h
Comment (apolcyn) [inline src/core/ext/gcp/metadata_query.h:48]: `result` might be better as a `StatusOr<std::string>` in order to propagate debug info. E.g., ` StatusOr<std::string>` would give callers the reason for query failure without enabling tracers and correlating logs.

Note this API is fine as is for the google-c2p resolver, since the c2p resolver internally handles query failures without propagating failures, but I don't know about other callers.

File: src/core/ext/gcp/metadata_query.cc
Comment (yashykt) [reply src/core/ext/gcp/metadata_query.cc]: removed

File: src/core/ext/gcp/metadata_query.cc
Comment (yashykt) [reply src/core/ext/gcp/metadata_query.cc]: done

File: src/core/ext/gcp/metadata_query.h
Comment (yashykt) [reply src/core/ext/gcp/metadata_query.h]: done

**Review by apolcyn — APPROVED** [review]
LGTM

File: src/core/BUILD
Comment (markdroth) [inline src/core/BUILD:5308]: Suggest calling this `gcp_metadata_query`.

File: src/core/ext/gcp/metadata_query.cc
Comment (markdroth) [inline src/core/ext/gcp/metadata_query.cc:123]: Is the body guaranteed to be nul-terminated and to contain no nul characters before the end?  If not, this isn't safe.  I think instead it should be:

```
result = std::string(self->response_.body, self->response_.body_length);
```

File: src/core/ext/gcp/metadata_query.h
Comment (markdroth) [inline src/core/ext/gcp/metadata_query.h:59]: I think we should use `grpc_error_handle` instead of `absl::Status` as the type for this parameter, since this is a C-style closure.  That will be a signal to us that this is something that needs to be cleaned up later.

File: src/cpp/ext/gcp/BUILD
Comment (markdroth) [inline src/cpp/ext/gcp/BUILD:134]: Should this be in `public_hdrs`?

File: src/cpp/ext/gcp/environment_autodetect.h
Comment (markdroth) [inline src/cpp/ext/gcp/environment_autodetect.h:72]: I don't think `Create()` should return the instance anymore, since it's just for initializing the global singleton.  Instead, it should set a static data member (or just a global in the .cc file), and `Get()` should directly return that global instead of calling `Create()`.

File: test/cpp/ext/gcp/environment_autodetect_test.cc
Comment (markdroth) [inline test/cpp/ext/gcp/environment_autodetect_test.cc:62]: Why is this commented out?

File: test/cpp/ext/gcp/environment_autodetect_test.cc
Comment (markdroth) [inline test/cpp/ext/gcp/environment_autodetect_test.cc:130]: Same question here.

**Review by markdroth — COMMENTED** [review]
I like this structure much better!

File: test/cpp/ext/gcp/environment_autodetect_test.cc
Comment (yashykt) [reply test/cpp/ext/gcp/environment_autodetect_test.cc:130]: This is a problem of our testing environments. I can't be sure whether we'd be running on an environment that has MetadataServer or not, so unless I mock the MetadataServer, I can not get a stable result here across environments.

File: test/cpp/ext/gcp/environment_autodetect_test.cc
Comment (yashykt) [reply test/cpp/ext/gcp/environment_autodetect_test.cc:62]: This is a problem of our testing environments. I can't be sure whether we'd be running on an environment that has MetadataServer or not, so unless I mock the MetadataServer, I can not get a stable result here across environments.

File: src/core/ext/gcp/metadata_query.cc
Comment (yashykt) [reply src/core/ext/gcp/metadata_query.cc]: Done. Good catch! Thanks!

File: src/core/ext/gcp/metadata_query.h
Comment (yashykt) [reply src/core/ext/gcp/metadata_query.h]: because we want c-style closures to go away? Done

File: src/cpp/ext/gcp/BUILD
Comment (yashykt) [reply src/cpp/ext/gcp/BUILD]: ah, yeah, removed

File: src/cpp/ext/gcp/environment_autodetect.h
Comment (yashykt) [reply src/cpp/ext/gcp/environment_autodetect.h]: Done

**Review by markdroth — APPROVED** [review]

**Pull request merged**
